File: backend/src/main/java/com/securebank/util/PasswordValidator.java
package com.securebank.util;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

import java.util.regex.Pattern;

public class PasswordValidator implements ConstraintValidator<ValidPassword, String> {
    
    // Critères de mot de passe fort
    private static final int MIN_LENGTH = 12;
    private static final Pattern UPPERCASE_PATTERN = Pattern.compile(".*[A-Z].*");
    private static final Pattern LOWERCASE_PATTERN = Pattern.compile(".*[a-z].*");
    private static final Pattern DIGIT_PATTERN = Pattern.compile(".*\\d.*");
    private static final Pattern SPECIAL_CHAR_PATTERN = Pattern.compile(".*[!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>\\/?].*");
    
    @Override
    public void initialize(ValidPassword constraintAnnotation) {
        // Initialisation si nécessaire
    }
    
    @Override
    public boolean isValid(String password, ConstraintValidatorContext context) {
        if (password == null) {
            return false;
        }
        
        boolean isValid = true;
        context.disableDefaultConstraintViolation();
        
        // Vérifier la longueur minimale
        if (password.length() < MIN_LENGTH) {
            context.buildConstraintViolationWithTemplate(
                "Le mot de passe doit contenir au moins " + MIN_LENGTH + " caractères"
            ).addConstraintViolation();
            isValid = false;
        }
        
        // Vérifier la présence de majuscules
        if (!UPPERCASE_PATTERN.matcher(password).matches()) {
            context.buildConstraintViolationWithTemplate(
                "Le mot de passe doit contenir au moins une lettre majuscule"
            ).addConstraintViolation();
            isValid = false;
        }
        
        // Vérifier la présence de minuscules
        if (!LOWERCASE_PATTERN.matcher(password).matches()) {
            context.buildConstraintViolationWithTemplate(
                "Le mot de passe doit contenir au moins une lettre minuscule"
            ).addConstraintViolation();
            isValid = false;
        }
        
        // Vérifier la présence de chiffres
        if (!DIGIT_PATTERN.matcher(password).matches()) {
            context.buildConstraintViolationWithTemplate(
                "Le mot de passe doit contenir au moins un chiffre"
            ).addConstraintViolation();
            isValid = false;
        }
        
        // Vérifier la présence de caractères spéciaux
        if (!SPECIAL_CHAR_PATTERN.matcher(password).matches()) {
            context.buildConstraintViolationWithTemplate(
                "Le mot de passe doit contenir au moins un caractère spécial (!@#$%^&*()_+-=[]{}|;':\"\\,.<>/?)"
            ).addConstraintViolation();
            isValid = false;
        }
        
        return isValid;
    }
}

File: frontend/src/utils/passwordValidator.js
export const validatePassword = (password) => {
  const errors = [];
  
  // Critères de validation
  const minLength = 12;
  const hasUppercase = /[A-Z]/.test(password);
  const hasLowercase = /[a-z]/.test(password);
  const hasDigit = /\d/.test(password);
  const hasSpecialChar = /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password);
  
  // Vérifications
  if (password.length < minLength) {
    errors.push(`Au moins ${minLength} caractères`);
  }
  
  if (!hasUppercase) {
    errors.push('Au moins une lettre majuscule');
  }
  
  if (!hasLowercase) {
    errors.push('Au moins une lettre minuscule');
  }
  
  if (!hasDigit) {
    errors.push('Au moins un chiffre');
  }
  
  if (!hasSpecialChar) {
    errors.push('Au moins un caractère spécial (!@#$%^&*()_+-=[]{}|;\':"\\,.<>/?)');
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
};

export const getPasswordStrength = (password) => {
  const validation = validatePassword(password);
  
  if (validation.isValid) {
    return 'strong';
  }
  
  if (password.length >= 8 && validation.errors.length <= 2) {
    return 'medium';
  }
  
  return 'weak';
};